src/report/report.c

```diff
@@ -0,0 +1,225 @@
+
+typedef struct Context {
+        int n_ref;
+} Context;
+
+static int metrics_on_query_reply(
+                sd_varlink *link,
+                sd_json_variant *parameters,
+                const char *error_id,
+                sd_varlink_reply_flags_t flags,
+                void *userdata) {
+        assert(link);
+
+        Context* context = ASSERT_PTR(userdata);
+
+        if (error_id) {
+                if (streq(error_id, SD_VARLINK_ERROR_DISCONNECTED))
+                        log_info("Varlink disconnected");
+                else if (streq(error_id, SD_VARLINK_ERROR_TIMEOUT))
+                        log_info("Varlink timed out");
+                else
+                        log_error("Varlink error: %s", error_id);
+
+                goto finish;
+        }
+
+        sd_json_variant_dump(parameters, SD_JSON_FORMAT_PRETTY_AUTO|SD_JSON_FORMAT_COLOR_AUTO, stdout, NULL);
+
+        fflush(stdout);
+
+finish:
+        if (!FLAGS_SET(flags, SD_VARLINK_REPLY_CONTINUES) && (--context->n_ref == 0))
+                (void) sd_event_exit(ASSERT_PTR(sd_varlink_get_event(link)), EXIT_SUCCESS);
+
+        return 0;
+}
+
+static int metrics_call(const char *path, sd_event *event, sd_varlink **ret, Context *context) {
+        _cleanup_(sd_varlink_unrefp) sd_varlink *vl = NULL;
+        int r;
+
+        assert(path);
+        assert(event);
+        assert(ret);
+        assert(context);
+
+        r = sd_varlink_connect_address(&vl, path);
+        if (r < 0)
+                return log_debug_errno(r, "Unable to connect to %s: %m", path);
+
+        (void) sd_varlink_set_userdata(vl, context);
+
+        r = sd_varlink_set_relative_timeout(vl, TIMEOUT_USEC);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set varlink timeout: %m");
+
+        r = sd_varlink_attach_event(vl, event, SD_EVENT_PRIORITY_NORMAL);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to attach varlink connection to event loop: %m");
+
+        r = sd_varlink_bind_reply(vl, metrics_on_query_reply);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to bind reply callback: %m");
+
+        r = sd_varlink_observe(vl, "io.systemd.Metrics.List", /* parameters= */ NULL);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to issue io.systemd.Metrics.List call: %m");
+
+        *ret = TAKE_PTR(vl);
+
+        return 0;
+}
+
+static void sd_varlink_unref_many(sd_varlink **array, size_t n) {
+        assert(array);
+
+        FOREACH_ARRAY(v, array, n)
+                sd_varlink_unref(*v);
+
+        free(array);
+}
+
+static int metrics_start_query(void) {
+        _cleanup_closedir_ DIR *d = NULL;
+        _cleanup_(sd_event_unrefp) sd_event *event = NULL;
+        int r;
+
+        d = opendir(RUN_SYSTEMD_METRICS_PATH);
+        if (!d) {
+                if (errno == ENOENT)
+                        return log_error_errno(ENOENT, "No metrics sockets found: %m");
```

Comment:
SYNTHETIC_ERRNO()

but i don#t think this should be an error. a notice at best, and this should be a success error code.